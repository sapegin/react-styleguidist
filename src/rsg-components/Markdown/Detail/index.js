export { default } from 'rsg-components/Markdown/Detail/DetailRenderer';
export { default as DetailSummary } from 'rsg-components/Markdown/Detail/DetailSummaryRenderer';
