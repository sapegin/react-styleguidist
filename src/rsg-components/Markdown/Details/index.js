export { default } from 'rsg-components/Markdown/Details/DetailsRenderer';
