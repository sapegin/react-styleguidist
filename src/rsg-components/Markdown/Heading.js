import React from 'react';
import PropTypes from 'prop-types';

import Styled from 'rsg-components/Styled';

import HeadingRenderer from 'rsg-components/Heading/HeadingRenderer';

function MarkdownHeading({ classes, depth, children }) {
	return (
		<div className={classes.spacing}>
			<HeadingRenderer depth={depth}>{children}</HeadingRenderer>
		</div>
	);
}

const styles = ({ space }) => ({
	spacing: {
		marginTop: 0,
		marginBottom: space[2],
	},
});

MarkdownHeading.propTypes = {
	classes: PropTypes.object.isRequired,
	depth: PropTypes.oneOf([1, 2, 3, 4, 5, 6]).isRequired,
	children: PropTypes.node,
};

export default Styled(styles)(MarkdownHeading);
