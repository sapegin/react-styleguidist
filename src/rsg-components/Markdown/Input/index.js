export { default } from 'rsg-components/Markdown/Input/InputRenderer';
