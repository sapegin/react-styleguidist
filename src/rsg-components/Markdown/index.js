import { asArrayMarkdown } from './Markdown';

export { default, asArrayMarkdown } from './Markdown.js';

