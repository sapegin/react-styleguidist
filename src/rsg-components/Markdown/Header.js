import React from 'react';
import PropTypes from 'prop-types';

import Styled from 'rsg-components/Styled';
import { styles as paraStyles } from 'rsg-components/Para';

const styles = ({ space, color, fontSize, fontFamily }) => ({
	para: paraStyles({ space, color, fontFamily }).para,
	h1: {
		composes: '$para',
		fontSize: fontSize.h1,
		fontWeight: 'normal',
	},
	h2: {
		composes: '$para',
		fontSize: fontSize.h2,
		fontWeight: 'normal',
	},
	h3: {
		composes: '$para',
		fontSize: fontSize.h3,
		fontWeight: 'normal',
	},
	h4: {
		composes: '$para',
		fontSize: fontSize.h4,
		fontWeight: 'normal',
	},
	h5: {
		composes: '$para',
		fontSize: fontSize.h5,
		fontWeight: 'normal',
	},
	h6: {
		composes: '$para',
		fontSize: fontSize.h6,
		fontWeight: 'normal',
		fontStyle: 'italic',
	},
});

function Header({ level, classes, children, ...props }) {
	return React.createElement(
		level,
		{
			...props,
			className: classes[level],
		},
		children
	);
}

Header.propTypes = {
	level: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
	classes: PropTypes.object.isRequired,
	children: PropTypes.node,
};

const StyledHeader = Styled(styles)(Header);

export default {
	h1: {
		component: StyledHeader,
		props: {
			level: 'h1',
		},
	},
	h2: {
		component: StyledHeader,
		props: {
			level: 'h2',
		},
	},
	h3: {
		component: StyledHeader,
		props: {
			level: 'h3',
		},
	},
	h4: {
		component: StyledHeader,
		props: {
			level: 'h4',
		},
	},
	h5: {
		component: StyledHeader,
		props: {
			level: 'h5',
		},
	},
	h6: {
		component: StyledHeader,
		props: {
			level: 'h6',
		},
	},
};
