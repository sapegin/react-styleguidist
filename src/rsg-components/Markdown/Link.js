import React from 'react';
import PropTypes from 'prop-types';

import Styled from 'rsg-components/Styled';
import { styles as linkStyles } from 'rsg-components/Link';

const styles = ({ color }) => ({
	a: linkStyles({ color }).link,
});

function Link({ classes, children, ...props }) {
	return (
		<a {...props} className={classes.a}>
			{children}
		</a>
	);
}

Link.propTypes = {
	classes: PropTypes.object.isRequired,
	children: PropTypes.node,
};

export default {
	a: {
		component: Styled(styles)(Link),
	},
};
