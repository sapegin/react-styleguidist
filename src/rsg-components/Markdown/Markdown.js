import React from 'react';
import PropTypes from 'prop-types';
import { compiler } from 'markdown-to-jsx';
import mapValues from 'lodash/mapValues';
// import memoize from 'lodash/memoize';
import Styled from 'rsg-components/Styled';
import Link from 'rsg-components/Link';
import { styles as paraStyles } from 'rsg-components/Para';
import MarkdownHeading from 'rsg-components/Markdown/MarkdownHeading';

// We’re explicitly specifying Webpack loaders here so we could skip specifying them in Webpack configuration.
// That way we could avoid clashes between our loaders and user loaders.
// eslint-disable-next-line import/no-unresolved
require('!!../../../loaders/style-loader!../../../loaders/css-loader!highlight.js/styles/tomorrow.css');

// Temporary disable memoization to fix: https://github.com/styleguidist/react-styleguidist/issues/348
// TODO: Remove after merge: https://github.com/probablyup/markdown-to-jsx/pull/96
const memoize = a => a;

// Code blocks with server-side syntax highlight
function Code({ children, className }) {
	const isHighlighted = className && className.indexOf('lang-') !== -1;
	if (isHighlighted) {
		return <code className={className} dangerouslySetInnerHTML={{ __html: children }} />;
	}
	return <code className={className}>{children}</code>;
}
Code.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
};

// Custom CSS classes for each tag: <em> → <em className={s.em}> + custom components
const getBaseOverrides = memoize(classes => {
	const styleOverrides = mapValues(classes, value => ({
		props: {
			className: value,
		},
	}));

	return {
		...styleOverrides,
		a: {
			component: Link,
		},
		h1: {
			component: MarkdownHeading,
			props: {
				level: 1,
			},
		},
		h2: {
			component: MarkdownHeading,
			props: {
				level: 2,
			},
		},
		h3: {
			component: MarkdownHeading,
			props: {
				level: 3,
			},
		},
		h4: {
			component: MarkdownHeading,
			props: {
				level: 4,
			},
		},
		h5: {
			component: MarkdownHeading,
			props: {
				level: 5,
			},
		},
		h6: {
			component: MarkdownHeading,
			props: {
				level: 6,
			},
		},
		code: {
			component: Code,
			props: {
				className: classes.code,
			},
		},
	};
}, () => 'getBaseOverrides');

// Inline mode: replace <p> (usual root component) with <span>
const getInlineOverrides = memoize(classes => {
	const overrides = getBaseOverrides(classes);

	return {
		...overrides,
		p: {
			component: 'span',
			props: {
				className: classes.base,
			},
		},
	};
}, () => 'getInlineOverrides');

const styles = ({ space, fontFamily, fontSize, color, borderRadius }) => ({
	base: {
		color: color.base,
		fontFamily: fontFamily.base,
		fontSize: 'inherit',
	},
	para: paraStyles({ space, color, fontFamily }).para,
	p: {
		composes: '$para',
	},
	ul: {
		composes: '$para',
		paddingLeft: space[3],
	},
	ol: {
		composes: '$para',
		listStyleType: 'decimal',
		paddingLeft: space[3],
	},
	li: {
		composes: '$base',
		listStyleType: 'inherit',
	},
	input: {
		color: color.base,
		display: 'inline-block',
		margin: [[0, '0.35em', '0.25em', '-1.2em']],
		verticalAlign: 'middle',
	},
	blockquote: {
		composes: '$para',
		fontSize: fontSize.base,
		margin: [[space[2], space[4]]],
		padding: 0,
	},
	hr: {
		composes: '$para',
		borderWidth: [[0, 0, 1, 0]],
		borderColor: color.border,
		borderStyle: 'solid',
	},
	em: {
		composes: '$base',
		fontStyle: 'italic',
	},
	strong: {
		composes: '$base',
		fontWeight: 'bold',
	},
	code: {
		fontFamily: fontFamily.monospace,
		fontSize: 'inherit',
		color: 'inherit',
		background: 'transparent',
		whiteSpace: 'inherit',
	},
	pre: {
		composes: '$para',
		backgroundColor: color.codeBackground,
		border: [[1, color.border, 'solid']],
		padding: [[space[1], space[2]]],
		fontSize: fontSize.small,
		borderRadius,
		whiteSpace: 'pre',
	},
	table: {
		composes: '$para',
		borderCollapse: 'collapse',
	},
	thead: {
		composes: '$hr',
	},
	tbody: {},
	td: {
		fontFamily: fontFamily.base,
		padding: [[space[0], space[2], space[0], 0]],
		fontSize: fontSize.base,
	},
	th: {
		composes: '$td',
		fontWeight: 'bold',
	},
	tr: {},
});

function Markdown({ classes, text, inline }) {
	const overrides = inline ? getInlineOverrides(classes) : getBaseOverrides(classes);
	return compiler(text, { overrides });
}

Markdown.propTypes = {
	classes: PropTypes.object.isRequired,
	text: PropTypes.string.isRequired,
	inline: PropTypes.bool,
};

export default Styled(styles)(Markdown);
