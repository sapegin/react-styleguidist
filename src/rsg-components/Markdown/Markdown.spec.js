import React from 'react';
import { html } from 'cheerio';
import Markdown from './Markdown';

describe('Markdown', () => {
	it('should render links', () => {
		const markdown = 'a [link](http://test.com)';

		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render Markdown with custom CSS classes', () => {
		const markdown = `
# Header

Text with *some* **formatting** and a [link](/foo).

![Image](/bar.png)`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render Markdown in a p tag even for one paragraph', () => {
		const actual = render(<Markdown text="pizza" />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render Markdown in span in inline mode', () => {
		const markdown = 'Hello *world*!';
		const actual = render(<Markdown text={markdown} inline />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render headings correctly', () => {
		const markdown = `
# one
## two
### three
#### four
##### five
###### six
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render unordered lists correctly', () => {
		const markdown = `
* list
* item
* three
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render ordered lists correctly', () => {
		const markdown = `
1. list
1. item
1. three
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render check-lists correctly', () => {
		const markdown = `
* [ ] list 1
* [ ] list 2
* [x] list 3
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render mixed nested lists correctly', () => {
		const markdown = `
* list 1
* list 2
  1. Sub-list
  1. Sub-list
  1. Sub-list
* list 3
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render code blocks without escaping', () => {
		const markdown = `
\`\`\`html
<foo></foo>
\`\`\`
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render inline code with escaping', () => {
		const markdown = 'Foo `<bar>` baz';

		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render a blockquote', () => {
		const markdown = `
> This is a blockquote.
> And this is a second line.
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});

	it('should render a table', () => {
		const markdown = `
| heading 1 | heading 2 |
| --------- | --------- |
| foo		| bar		|
| more foo	| more bar	|
`;
		const actual = render(<Markdown text={markdown} />);

		expect(html(actual)).toMatchSnapshot();
	});
});
