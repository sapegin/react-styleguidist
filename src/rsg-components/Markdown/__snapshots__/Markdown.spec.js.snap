// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown should render Markdown in span in inline mode 1`] = `"<span class=\\"rsg--base-0\\">Hello <em class=\\"rsg--em-9 rsg--base-0\\">world</em>!</span>"`;

exports[`Markdown should render Markdown with custom CSS classes 1`] = `"<div><div class=\\"rsg--spacing-21\\"><h1 class=\\"rsg--heading-22 rsg--heading1-23\\">Header</h1></div><p class=\\"rsg--p-2 rsg--para-1\\">Text with <em class=\\"rsg--em-9 rsg--base-0\\">some</em> <strong class=\\"rsg--strong-10 rsg--base-0\\">formatting</strong> and a <a href=\\"/foo\\" class=\\"rsg--link-20\\">link</a>.</p><p class=\\"rsg--p-2 rsg--para-1\\"><img alt=\\"Image\\" src=\\"/bar.png\\"></p></div>"`;

exports[`Markdown should render check-lists correctly 1`] = `"<ul class=\\"rsg--ul-3 rsg--para-1\\"><li class=\\"rsg--li-5 rsg--base-0\\"><input type=\\"checkbox\\" class=\\"rsg--input-6\\" readonly>list 1</li><li class=\\"rsg--li-5 rsg--base-0\\"><input type=\\"checkbox\\" class=\\"rsg--input-6\\" readonly>list 2</li><li class=\\"rsg--li-5 rsg--base-0\\"><input type=\\"checkbox\\" class=\\"rsg--input-6\\" checked readonly>list 3</li></ul>"`;

exports[`Markdown should render code blocks without escaping 1`] = `"<pre class=\\"rsg--pre-12 rsg--para-1\\"><code class=\\"rsg--code-11 lang-html\\"><foo></foo></code></pre>"`;

exports[`Markdown should render headings correctly 1`] = `"<div><div class=\\"rsg--spacing-21\\"><h1 class=\\"rsg--heading-22 rsg--heading1-23\\">one</h1></div><div class=\\"rsg--spacing-21\\"><h2 class=\\"rsg--heading-22 rsg--heading2-24\\">two</h2></div><div class=\\"rsg--spacing-21\\"><h3 class=\\"rsg--heading-22 rsg--heading3-25\\">three</h3></div><div class=\\"rsg--spacing-21\\"><h4 class=\\"rsg--heading-22 rsg--heading4-26\\">four</h4></div><div class=\\"rsg--spacing-21\\"><h5 class=\\"rsg--heading-22 rsg--heading5-27\\">five</h5></div><div class=\\"rsg--spacing-21\\"><h6 class=\\"rsg--heading-22 rsg--heading6-28\\">six</h6></div></div>"`;

exports[`Markdown should render inline code with escaping 1`] = `"<p class=\\"rsg--p-2 rsg--para-1\\">Foo <code class=\\"rsg--code-11\\">&lt;bar&gt;</code> baz</p>"`;

exports[`Markdown should render links 1`] = `"<p class=\\"rsg--p-2 rsg--para-1\\">a <a href=\\"http://test.com\\" class=\\"rsg--link-20\\">link</a></p>"`;

exports[`Markdown should render mixed nested lists correctly 1`] = `"<ul class=\\"rsg--ul-3 rsg--para-1\\"><li class=\\"rsg--li-5 rsg--base-0\\">list 1</li><li class=\\"rsg--li-5 rsg--base-0\\"><p class=\\"rsg--p-2 rsg--para-1\\">list 2</p><ol class=\\"rsg--ol-4 rsg--para-1\\" start=\\"1\\"><li class=\\"rsg--li-5 rsg--base-0\\">Sub-list</li><li class=\\"rsg--li-5 rsg--base-0\\">Sub-list</li><li class=\\"rsg--li-5 rsg--base-0\\">Sub-list</li></ol></li><li class=\\"rsg--li-5 rsg--base-0\\">list 3</li></ul>"`;

exports[`Markdown should render ordered lists correctly 1`] = `"<ol class=\\"rsg--ol-4 rsg--para-1\\" start=\\"1\\"><li class=\\"rsg--li-5 rsg--base-0\\">list</li><li class=\\"rsg--li-5 rsg--base-0\\">item</li><li class=\\"rsg--li-5 rsg--base-0\\">three</li></ol>"`;

exports[`Markdown should render unordered lists correctly 1`] = `"<ul class=\\"rsg--ul-3 rsg--para-1\\"><li class=\\"rsg--li-5 rsg--base-0\\">list</li><li class=\\"rsg--li-5 rsg--base-0\\">item</li><li class=\\"rsg--li-5 rsg--base-0\\">three</li></ul>"`;
